Updates to schema with required changes to source files, and new README.

dkerr64 opened the pull request:
Addresses issues #26 and #19

kovapatrik commented:
I don't really get why the AC/dehumidifer options must be optional. We can define default in the config scheme itself, so it doesn't have to be handled in the code.

dkerr64 commented:
> I don't really get why the AC/dehumidifer options must be optional. We can define defaults in the config scheme itself, so it doesn't have to be handled in the code.
> 

I removed all (most) of the default assignments from the schema.  If you define a default in the schema file then that entry is always written to the config file.  This means that an entry for a Dehumidifier will have settings set for itself plus for an air conditioner... every single one.  If this plugin expands to support additional devices then the config file will get very large very quickly.

dkerr64 commented:
> Also why would we allow an empty device config? The point of the config per device is that you know which IP got reserved for your device, and you can configure the device as you want. An empty device config simply doesn't make sense for me.

The plugin should function without the user having to make any entry for any device into the config file.  The plugin automatically discovers all devices on the network and configures them.  There is no need for anything in the config file, therefore the default should not require it.  It is just simpler for users.

kovapatrik commented:
Ah okay, both makes sense. My second thoughts about this;
- I think it's not a problem that the config file will be bigger. A normal user won't interact with the raw config file, only with the GUI for the plugin. The increasing of the size of the config file is also irrelevant in my opinion.
- I have an other plugin for Xiaomi devices, which has an autodiscover option too but it's not automatic: it appaers as a button in the config GUI. If a user clicks on it, it will discover all compatible devices on the network and the user can decide which device will be added. If a device is added it will be written in the config file.
- The empty configuration makes sense in your way, but if we intend to implement my last point, it is not needed. In my opinion it's a better flow to not autodiscover the devices but to do it manually.

dkerr64 commented:
Making auto-discover optional sounds like a reasonable enhancement.  But I'm definitely not in favor of populating the config file with settings that are unique to one device (e.g. air conditioner) into a section that is for another (e.g. dehumidifier) device.  That is just confusing.

And I wouldn't assume that everyone does configuration through the GUI.  I certainly directly view/edit the config file.  And I even have plugins that do not implement a GUI settings page and require that I edit the config file.

dkerr64 commented:
I've been busy this weekend, as you can tell, and I think I have the plugin where I want it to be (with this PR).  It should be pretty robust now and in a good place to allow adding additional devices.

Put another way, I don't have anything else that I want to add so won't be proactively adding things.  But I'm happy to work on anything you want to assign to me :-)

kovapatrik commented:
> Making auto-discover optional sounds like a reasonable enhancement.  But I'm definitely not in favor of populating the config file with settings that are unique to one device (e.g. air conditioner) into a section that is for another (e.g. dehumidifier) device.  That is just confusing.
> 
> 
> 
> And I wouldn't assume that everyone does configuration through the GUI.  I certainly directly view/edit the config file.  And I even have plugins that do not implement a GUI settings page and require that I edit the config file.

Oh, I didn't realise to this point that the settings for AC will be populated to a dehumidifer. I think (or I hope :D) there should be an option to handle this nicely in the config scheme.

An official (or approved) homebridge plugin MUST have a GUI for the config, that's why I would rely on it; an end user won't mess with the raw file.

dkerr64 commented:
I added placeholder values to the config schema. However it appears that placeholders on boolean type is being ignored. I have opened a [bug report at homebridge-config-ui-x](https://github.com/homebridge/homebridge-config-ui-x/issues/1594) and left the placeholders in here in hope that the bug is fixed

dkerr64 commented:
Also, is there a reason that you spelled the variable `fahrenHeit` with a capitalized letter H?

kovapatrik commented:
In other plugins I'm using it works as you described: all settings are listed for each device, and only the relevant ones will be used by the accessory. I think it's not a good solution though, the config won't be that transparent.

I like the placeholder solution, nice discovery! I hope the bug will get fixed soon.

I was writing everything in camel case and I'm using Copilot. I think the AI just suggested it and I didn't payed attention when I've accepted it.

dkerr64 commented:
Changing the capitalization in `fahrenHeit` will be a breaking change.  We could add code to accept either spelling, but I'm not sure it is worth it.  It has minor consequences affecting display only.  But I do think it should be corrected to `fahrenheit`.

kovapatrik commented:
True, but as I said before: the plugin isn't even in 'production' stage, all current users are beta testers. So there is currently no breaking change in my opinion. There are plugins, where there is a message popup after an update. We can warn the current users about breaking changes this way.

kovapatrik commented:
I think we should also implement the button based discovery in this update to the scheme would be finalized.
The [homebridge-miot](url) plugin config GUI looks like this: 
![image](https://github.com/kovapatrik/homebridge-midea-platform/assets/44952801/6352b3e6-4f01-4f83-bb80-1cb1e5c11e10)
I think it's much more transparent like this. What are your thoughts?

dkerr64 commented:
I think adding a button to the GUI for device discovery would be pretty cool.  I definitely like the idea.  But... it looks non-trivial. The homebridge-miot plugin has an awful lot of code to implement those buttons.  It is not clear to me yet how much of the existing discovery/midea-cloud code we can simply call from either the `server.js` file or javascript within `index.html` file that implement the customUI.  I would have to study it a lot more.

Let's open an issue/feature request to discuss/track.  But I would not hold up merging this PR for it.

kovapatrik commented:
Yeah, you are right. I will implement it in the future. By the way the homebridge-miot repo has all the device options for all device independent of the type. With the placeholder option the default ones won't get written in the config file, as you found out. If we define a default value for all fields, then we can handle the config settings as they MUST exists and we don't have to set defaults in the code itself.

kovapatrik commented:
Indexing the devices by their id also don't make much sense for me. The object you created will not persist and it doesn't have any usage as the current state of the code.

kovapatrik commented:
I just realized that the placeholder field won't do anything besides showing a value to the user, so defaults must be set in the code. Only problem with this placeholder solution is that enums or booleans values cannot be set by placeholders. I think this can be solved with the custom UI. I've started to implement it, and it's working so fine. I will add the handling of the placeholder values into it and it will be done.

dkerr64 commented:
> I just realized that the placeholder field won't do anything besides showing a value to the user, so defaults must be set in the code. Only problem with this placeholder solution is that enums or booleans values cannot be set by placeholders. I think this can be solved with the custom UI. I've started to implement it, and it's working so fine. I will add the handling of the placeholder values into it and it will be done.

Yes, I had the same thought... with custom UI it should be possible to mark the checkbox for a boolean true placeholder.  It may be more complicated to get the full functionality of placeholder to work.... it is only supposed to write a value to config if the user changes it from default, or if it already exists in the config.

I know you have a different view than I do, but I do not think any value should be written to the config file unless it is different from the default, it just clutters the config file.  And I strongly believe that code should always permit an undefined config setting and set it to a default.  I added that to the constructors of each class.

Great that you have made a start on custom UI.  In my reading over the last couple of days I decided that the examples in [homebridge/plugin-ui-utils](https://github.com/homebridge/plugin-ui-utils) would be best place to start from.  Nice and simple.  I was also thinking that server.js should be kept as small/simple as possible... calling into functions implemented in the main typescript source for the plugin (maybe a new file in core, e.g. MideaCustomUI.ts.

Thanks.

dkerr64 commented:
> Indexing the devices by their id also don't make much sense for me. The object you created will not persist and it doesn't have any usage as the current state of the code.

devicesById (and devicesByIP) are used in the finishedLaunching function and deviceDiscovered function.  It is a much better way of referencing a specific device and the conversion I do in the constructor is the [recommended method](https://developers.homebridge.io/#/config-schema#user-defined-keys). Except that I move the object into a new array rather than replace the existing one... because I want to be able to identify if any device was configured without either ID or IP address (which may be desirable if you want to have setting apply to all devices of the same type... not implemented yet of course).

dkerr64 commented:
> ..... handle the config settings as they MUST exists and we don't have to set defaults in the code itself.

I strongly disagree with this.  There MUST NOT be a requirement for all settings to exist in the config file.  The code MUST handle any situations where a config setting is undefined.  This is just good practice.

kovapatrik commented:
> > ..... handle the config settings as they MUST exists and we don't have to set defaults in the code itself.
> 
> I strongly disagree with this. There MUST NOT be a requirement for all settings to exist in the config file. The code MUST handle any situations where a config setting is undefined. This is just good practice.

No no, you misunderstood me, or maybe I wrote it badly. I totally agree with default settings should not be written to the config file. I just thought that values in the placeholder field can be used in code, so then we can be sure about they have a value. But my last comment denied this hypothesis.

Indexing the devices with their ID still doesn't make sense for me. With the button added, all devices will have their ID field filled (and it's a required field). I thought about what if the IP of the device changes. I would implement a flow like this:

1. If a device has it's IP field filled, then discover by the IP (still the ID field needs to be also filled)
2. If it's not, then the ID field is filled, so start a discovery at the start of the plugin, and get the IP of the device.

One problem remains: what if a device's IP changes during running? It can be detected I think so a discovery should start?

What are your thoughts about all this?

kovapatrik commented:
An other solution for the placeholder problem: we can define the optional variables' description such as the default is false. For example: the logRecoverableErrors should be true by default, so we can define the field in the config as the checkbox will disable it.

dkerr64 commented:
> Also discovering by IP seems kinda redundant for me this way.
> 
> One problem remains: what if a device's IP changes during running? It can be detected I think so a discovery should start?
> 
> What are your thoughts about all this?

I agree, once device discovery is done with a click in the UI, the existing discovery method is redundant... well, maybe, see below.

If a device IP changes while running then attempts to communicate with it will fail, and timeout process will trigger.  Just like a device getting unplugged.  But it will not (as currently coded) attempt a new discovery.  I am not sure we should even attempt that.  Of more concern would be if device IPs got swapped so all of a sudden a different type of device starts responding to messages.  Is that even possible?

What we can do is check the configured devices during plugin startup... do device discovery as we do now and check that it matches what is in the config file.  If it is different in any way (new device found, device not found, mismatch between IP / ID and Type, etc.) then we can handle that and issue a warning message to log asking user to re-do discovery in the Config UI.

dkerr64 commented:
> An other solution for the placeholder problem: we can define the optional variables' description such as the default is false. For example: the logRecoverableErrors should be true by default, so we can define the field in the config as the checkbox will disable it.

Yes, this might be the easiest way... for example change logRecoverableErrors to suppressRecoverableErrors.  Same for anything similar.

kovapatrik commented:
> > Also discovering by IP seems kinda redundant for me this way.
> > One problem remains: what if a device's IP changes during running? It can be detected I think so a discovery should start?
> > What are your thoughts about all this?
> 
> I agree, once device discovery is done with a click in the UI, the existing discovery method is redundant... well, maybe, see below.
> 
> If a device IP changes while running then attempts to communicate with it will fail, and timeout process will trigger. Just like a device getting unplugged. But it will not (as currently coded) attempt a new discovery. I am not sure we should even attempt that. Of more concern would be if device IPs got swapped so all of a sudden a different type of device starts responding to messages. Is that even possible?
> 
> What we can do is check the configured devices during plugin startup... do device discovery as we do now and check that it matches what is in the config file. If it is different in any way (new device found, device not found, mismatch between IP / ID and Type, etc.) then we can handle that and issue a warning message to log asking user to re-do discovery in the Config UI.

Yeah, I agree. Just not with the IP mismatch part. Discovery-wise it does not matter if the IP changes as long as we know the ID of the device. That's why I think we don't need discovery by IP, because a discovery will run every single time at startup.

dkerr64 commented:
> Yeah, I agree. Just not with the IP mismatch part. Discovery-wise it does not matter if the IP changes as long as we know the ID of the device. That's why I think we don't need discovery by IP, because a discovery will run every single time at startup.

I left in discovery by IP because early on you indicated that you had problems that required directly querying the device IP?  But if broadcast discovery is working for you then yes, take it out.  I don't know why broadcast would not work (unless someone configures their network switch to block broadcasting). I even took the extra step to find the subnets of each connected interface and broadcast to each (because 255.255.255.255 is sent only to the "first" interface).  So I would take it out.

The other thing I would do is remove userid/password/registered app from the config file.  Login to Midea cloud is only required to obtain token/key which can be done from the custom config UI.  So require these three values only to do the initial discovery and retrieval of token/key... user's browser can remember the userid/password for them and registered app can be selected from the three choices each time... no need to save them in the homebridge config.

kovapatrik commented:
Good point. Hopefully I can finish the custom UI thing in the next few days.

dkerr64 commented:
Found a bug.... in the Midea Dehumidifier!!!  But we need a work-around in our code...

I changed my router to force a specific IP address DHCP assignment to the dehumidifier.  When the lease expired on the old IP address (192.168.29.131), the router assigned the new IP address I selected (192.168.29.75).  As expected the plugin fails to get responses from status refresh or heartbeat.

However on restarting the plugin I would have expected it to work,  It did not...

```
[10/30/2023, 9:33:56 PM] [Midea] Start device discovery...
[10/30/2023, 9:33:56 PM] Homebridge v1.6.1 (HAP v0.11.1) (Midea) is running on port 41080.
[10/30/2023, 9:33:59 PM] [Midea] Sending discovery message to 192.168.17.255, try 1...
[10/30/2023, 9:33:59 PM] [Midea] Sending discovery message to 192.168.23.255, try 1...
[10/30/2023, 9:33:59 PM] [Midea] Sending discovery message to 192.168.12.255, try 1...
[10/30/2023, 9:33:59 PM] [Midea] Sending discovery message to 192.168.29.255, try 1...
[10/30/2023, 9:34:00 PM] [Midea] IP address mismatch: 192.168.29.131 != 192.168.29.75
[10/30/2023, 9:34:00 PM] [Midea] Discovered device: {"ip":"192.168.29.131","port":6444,"id":150633093209019,"model":"00000Q1B","sn":"000000P0000000Q1B88C291B363A0000","name":"net_a1_363A","type":161,"version":3}
[10/30/2023, 9:34:00 PM] [Midea] [Dehumidifier] Restoring existing accessory from cache: Dehumidifier
[10/30/2023, 9:34:00 PM] [Midea] [Dehumidifier] Device specific log recoverable errors is set to false
[10/30/2023, 9:34:00 PM] [Midea] [Dehumidifier] Cached device, using token/key from config file
[10/30/2023, 9:34:00 PM] [Midea] Connecting to device Dehumidifier (192.168.29.131:6444)...
[10/30/2023, 9:34:02 PM] [Midea] Sending discovery message to 192.168.17.255, try 2...
[10/30/2023, 9:34:02 PM] [Midea] Sending discovery message to 192.168.23.255, try 2...
[10/30/2023, 9:34:02 PM] [Midea] Sending discovery message to 192.168.12.255, try 2...
[10/30/2023, 9:34:02 PM] [Midea] Sending discovery message to 192.168.29.255, try 2...
[10/30/2023, 9:34:03 PM] [Midea] Socket error:
Error: connect EHOSTUNREACH 192.168.29.131:6444
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1495:16)
[10/30/2023, 9:34:03 PM] [Midea] [Dehumidifier] Error when connecting to device Dehumidifier (192.168.29.131:6444):
Error: connect EHOSTUNREACH 192.168.29.131:6444
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1495:16)
[10/30/2023, 9:34:03 PM] [Midea] [Dehumidifier] Starting network listener.
```

The reply from the broadcast message came from 192.168.29.75 as expected, but the content of the DeviceInfo message sent by the dehumidifier still held the old IP address... the device had not updated itself!  While we issue a warning message we did not handle it properly... we used the IP address in the message body when we should have used the IP address from which the reply came.  So I have [fixed that](https://github.com/kovapatrik/homebridge-midea-platform/pull/27/commits/f0c206860eabfd35924cc990cdf123f472c5b085)

Disconnecting power from the dehumidifier, then reconnecting causes the dehumidifier to notice that its IP address has changed.

kovapatrik commented:
Nice finding! Wonder why the creators of this protocol designed it this way... It's kinda weird for me.

kovapatrik commented:
I've created a PR to your branch. On my side it's working. I need your suggestion on visualizing the token/key values in the config UI, because as you can see, it's not looking great.

dkerr64 commented:
I merged it so I could test.  I did a quick Look at the code diff and may have some comments to add later.  In the meantime here are comments...

1. My device is discovered, but at the end of the process, before I "add" it, I get an error `[homebridge-midea-platform] Getting token and key with little-endian is not successful:` It appears twice.  Log copy/paste below.
2. I see what you mean by the token/key visualization.  I have two ideas, 1) decide how many characters to allow for each and then display say first 4-6 characters, dot dot dot, last 4-6 characters... `88d9b5...f2f14d` and 2) split it into two lines.
3. The discovery process is silent, even to debug log.  It would be good to log the "sending discovery message to..." for reassurance.  It would be good to try and push some sort of update to the GUI each loop, but that may not be possible.
4. The "single accessory" option isn't applicable to dehumidifier, I would prefer it to be with device specific section.

The log...
```
[10/31/2023, 7:46:14 PM] Homebridge v1.6.1 (HAP v0.11.1) (Homebridge 3CAB) is running on port 51517.
[10/31/2023, 7:46:51 PM] [Homebridge UI] [homebridge-midea-platform] Incoming Request: /getDefaults
[10/31/2023, 7:46:51 PM] [Homebridge UI] [homebridge-midea-platform] Incoming Request: /mergeToDefault
[10/31/2023, 7:47:02 PM] [Homebridge UI] [homebridge-midea-platform] Incoming Request: /login
[10/31/2023, 7:47:03 PM] [Homebridge UI] [homebridge-midea-platform] Incoming Request: /discover
[10/31/2023, 7:47:18 PM] [Homebridge UI] [homebridge-midea-platform] Getting token and key with little-endian is not successful:
Error: Token or key is missing!
    at UiServer.authenticate (/home/david/github/homebridge-midea-platform/homebridge-ui/server.js:100:13)
    at UiServer.getNewCredentials (/home/david/github/homebridge-midea-platform/homebridge-ui/server.js:87:20)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /home/david/github/homebridge-midea-platform/homebridge-ui/server.js:61:30
    at async Promise.all (index 0)
    at async Object./discover (/home/david/github/homebridge-midea-platform/homebridge-ui/server.js:49:24)
    at async UiServer.processRequest (/home/david/github/homebridge-midea-platform/node_modules/@homebridge/plugin-ui-utils/dist/server.js:75:30)
[10/31/2023, 7:47:19 PM] [Homebridge UI] [homebridge-midea-platform] Getting token and key with big-endian is not successful:
Error: Token or key is missing!
    at UiServer.authenticate (/home/david/github/homebridge-midea-platform/homebridge-ui/server.js:100:13)
    at UiServer.getNewCredentials (/home/david/github/homebridge-midea-platform/homebridge-ui/server.js:87:20)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /home/david/github/homebridge-midea-platform/homebridge-ui/server.js:61:30
    at async Promise.all (index 0)
    at async Object./discover (/home/david/github/homebridge-midea-platform/homebridge-ui/server.js:49:24)
    at async UiServer.processRequest (/home/david/github/homebridge-midea-platform/node_modules/@homebridge/plugin-ui-utils/dist/server.js:75:30)
[10/31/2023, 7:47:26 PM] [Homebridge UI] Changes to config.json saved.
[10/31/2023, 7:47:26 PM] [Homebridge UI] [homebridge-midea-platform] Terminating child process...
[10/31/2023, 7:47:26 PM] [Homebridge UI] [homebridge-midea-platform] Child process ended
[10/31/2023, 7:47:58 PM] [Homebridge UI] Changes to config.json saved.
```

Possibly more thoughts to come after I look at the code.  In the meantime once restarted the plugin does work for my device.

dkerr64 commented:
I found a way to add console logging.  Works fairly well, but occasionally the timestamp/prefix is missing. Don't know why...

```
[11/2/2023, 10:34:33 PM] [Homebridge UI] [homebridge-midea-platform] Start device discovery...
[11/2/2023, 10:34:33 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.17.255, try 1...
[11/2/2023, 10:34:33 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.23.255, try 1...
[11/2/2023, 10:34:33 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.12.255, try 1...
[11/2/2023, 10:34:33 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.29.255, try 1...
[11/2/2023, 10:34:33 PM] [Homebridge UI] [homebridge-midea-platform] Discovered device: {"ip":"192.168.29.75","port":6444,"id":150633093209019,"model":"00000Q1B","sn":"000000P0000000Q1B88C291B363A0000","name":"net_a1_363A","type":161,"version":3}
[11/2/2023, 10:34:36 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.17.255, try 2...
[11/2/2023, 10:34:36 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.23.255, try 2...
Sending discovery message to 192.168.12.255, try 2...
Sending discovery message to 192.168.29.255, try 2...
[11/2/2023, 10:34:39 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.17.255, try 3...
[11/2/2023, 10:34:39 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.23.255, try 3...
[11/2/2023, 10:34:39 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.12.255, try 3...
[11/2/2023, 10:34:39 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.29.255, try 3...
[11/2/2023, 10:34:42 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.17.255, try 4...
[11/2/2023, 10:34:42 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.23.255, try 4...
Sending discovery message to 192.168.12.255, try 4...
[11/2/2023, 10:34:42 PM] [Homebridge UI] [homebridge-midea-platform] Sending discovery message to 192.168.29.255, try 4...
[11/2/2023, 10:34:45 PM] [Homebridge UI] [homebridge-midea-platform] Device discovery complete after 4 network broadcasts.
```

In above example, three messages are missing the prefix, but it varies each run. However, it is better than nothing and will facilitate debugging.

dkerr64 commented:
I also discovered that the `setInterval()` function doesn't call the first instance of the function until after the timeout.  So there is a 3 second delay before sending the first broadcasts.  I fixed that too.

dkerr64 commented:
Realized that the _Getting token and key with little-endian is not successful_ message is "expected".  I was alarmed by it coming out in the log in error red color!!!  I have changed the log level to debug but also know that the `Logger()` stub I added does not check to see if debug logging is enabled or not, so it will always be printed to log.

I have to figure out how to detect if homebridge debug logging is enabled or not.  Too late in the evening to start that though.

dkerr64 commented:
We should also consider obtaining the token/key for each device as it is discovered.  Right now the code waits until the end of the full discovery process (12 seconds with 4 trys) before requesting the credentials.

dkerr64 commented:
I found quite a few bugs in the authenticate / getCredentials part of the code so I fixed all those.  And improved debug/log support.  I cannot find any way (at least not yet) to determine whether homebridge was started with `-D` for debug, so I have added a uiDebug field to the config and use that.  This is in fact exactly what homebridge-config-ui-x does as well, but I decided not to use its debug flag and instead use our own.  I think device discovery should be solid now.

dkerr64 commented:
Oops, sorry, hit close by mistake, reopened.

I have started looking at index.html and added some console logging so I can see what is going on client-side with the browser tools.  I have updated the visual for token/key to truncate them and put both in the same column.  For me that makes it fit in the available space.  Please check if it works for you.

Also, I get a error both client side and server side...
```
[Error] Failed to load resource: the server responded with a status of 404 (Not Found) (bootstrap.min.css.map, line 0)
[Error] Failed to load resource: the server responded with a status of 404 (Not Found) (bootstrap.min.js.map, line 0)
```
Do you see that as well?  The config UI is working, but it's a little disconcerting.

And I think we should allow "updating" of a device if it already exists. Every time you do a discover you get a new token/key.  I think it would be nice to update those in the config.json if the user wants.

That's it for now, probably more to come.

kovapatrik commented:
> Oops, sorry, hit close by mistake, reopened.
> 
> I have started looking at index.html and added some console logging so I can see what is going on client-side with the browser tools. I have updated the visual for token/key to truncate them and put both in the same column. For me that makes it fit in the available space. Please check if it works for you.
> 
> Also, I get a error both client side and server side...
> 
> ```
> [Error] Failed to load resource: the server responded with a status of 404 (Not Found) (bootstrap.min.css.map, line 0)
> [Error] Failed to load resource: the server responded with a status of 404 (Not Found) (bootstrap.min.js.map, line 0)
> ```
> 
> Do you see that as well? The config UI is working, but it's a little disconcerting.
> 
> And I think we should allow "updating" of a device if it already exists. Every time you do a discover you get a new token/key. I think it would be nice to update those in the config.json if the user wants.
> 
> That's it for now, probably more to come.

You can ignore those error. Those files are for debugging purposes, but we don't want to debug bootstrap itself.

kovapatrik commented:
The update button is a must with the solution you implemented for visualizing the token/key. Maybe it should be in an other color than purple to easily seperate the add and update buttons.

dkerr64 commented:
I found a bug in `filterOutDefaults`, need to handle the defaults object being undefined.  I've also added ability to update existing devices with new token/key, and added uiDebug to config schema.

Other thoughts/questions...

1. Do you know if it is possible to move the custom UI discover devices part to below the Optional Settings section?
2. Is it okay to move the Single Accessory setting to inside the AC settings?  I think it should be device type specific and not for all devices.
3. We should add "name"  to the config, and consider whether changing "platform" to "midea-platform"

dkerr64 commented:
> The update button is a must with the solution you implemented for visualizing the token/key. Maybe it should be in an other color than purple to easily seperate the add and update buttons.

Yes, color probably could change to amber or something.  I only have the one device so big gap in anything I do is testing for when multiple devices are discovered.  I can fake out multiple devices in the config.json but that doesn't test discovery and how it would look in the GUI. 

We need someone with multiple devices to help test.  I also have to test installation from scratch (ie, no entry at all in the config.json file yet)

kovapatrik commented:
> I found a bug in `filterOutDefaults`, need to handle the defaults object being undefined.  I've also added ability to update existing devices with new token/key, and added uiDebug to config schema.
> 
> 
> 
> Other thoughts/questions...
> 
> 
> 
> 1. Do you know if it is possible to move the custom UI discover devices part to below the Optional Settings section?
> 
> 2. Is it okay to move the Single Accessory setting to inside the AC settings?  I think it should be device type specific and not for all devices.
> 
> 3. We should add "name"  to the config, and consider whether changing "platform" to "midea-platform"
> 
> 
> 
> 

1. I think that's not possible. I didn't find anything about this in the docs.
2. The field is for merging accessories when multiple services created for a device. I think it's much common if a device has multiple peripheries (like temperature sensors), but maybe the field can be completely deleted. I added it because when I added a temperature sensor for my AC, the two services couldn't be seperated in the Home app for some reason.
3. The platform field in the config must be names as platform and it's value must match with the value in the settings.ts file.

kovapatrik commented:
> > The update button is a must with the solution you implemented for visualizing the token/key. Maybe it should be in an other color than purple to easily seperate the add and update buttons.
> 
> 
> 
> Yes, color probably could change to amber or something.  I only have the one device so big gap in anything I do is testing for when multiple devices are discovered.  I can fake out multiple devices in the config.json but that doesn't test discovery and how it would look in the GUI. 
> 
> 
> 
> We need someone with multiple devices to help test.  I also have to test installation from scratch (ie, no entry at all in the config.json file yet)

I have 4 air conditioners, and I also tested it from scratch and it worked.

dkerr64 commented:
> 1. I think that's not possible. I didn't find anything about this in the docs.
> 2. The field is for merging accessories when multiple services created for a device. I think it's much common if a device has multiple peripheries (like temperature sensors), but maybe the field can be completely deleted. I added it because when I added a temperature sensor for my AC, the two services couldn't be seperated in the Home app for some reason.
> 3. The platform field in the config must be named as 'platform' and it's value must match with the value in the settings.ts file.

1. Yup, I could not find anything either.
2. Yes I understand it can be desirable to combine multiple devices into one.  My only point is that this is very device specific.
3. When you search for plugins using the keyword midea several are found.  At least one of them uses the platform name 'midea'.  I think platform needs to be unique.  If someone already has this other plugins installed then I expect there will be problems when our plugin is added.  Hence my suggestion that we change it to 'midea-platform'

kovapatrik commented:
In the current state of the plugin, yes it is dependent on the device, but as I said I think it's much more common that a device has multiple peripheries. 

Oh, I misunderstood the platform thing a bit. We can change the name of the platform but I think there will be no case when a user has multiple midea related plugins installed.

dkerr64 commented:
I changed the platform name to `midea-platform` because I'm sure it needs to be unique and cannot be same as an existing plugin.  I also changed the update button in the GUI to be outline class which looks quite good and distinguishes it from the add button.

And I have updated the README to match current code.

While doing that I noted that we are not saving the IP address any more.  I had written code that would register an accessory in Homebridge even if the device was offline.  That looks like it was removed.  I want to put that back because it improves resiliency, for example recovery from a power outage.

kovapatrik commented:
The only thing that must be unique is the name of the npm package.

I dont't think that's a good idea. The IP address of the devices could change. When there is a power outage, the Homebridge instance itself will be also down, so the discovery process will start.

dkerr64 commented:
> I dont't think that's a good idea. The IP address of the devices could change. When there is a power outage, the Homebridge instance itself will be also down, so the discovery process will start after the power is back.

Yes, you are right, there is a better way.  See commit I just pushed... detect if a device in config.json was not discovered and then keep checking every 60 seconds until it is found.

This will allow for a device being offline when the plugin initializes.  When the device does come online, it will be found and added.  There was existing code to handle case where device is online when plugin initializes but then goes offline for a period and then comes back.

kovapatrik commented on src/platform.ts:
```diff
@@ -46,13 +50,22 @@ export class MideaPlatform implements DynamicPlatformPlugin {
     // Add default config values
     this.platformConfig = defaultsDeep(config, defaultConfig);
+    // Enforce min/max values
+    this.platformConfig.refreshInterval = Math.max(0, Math.min(this.platformConfig.refreshInterval ?? 30, 86400));
```
It's not neccessary to handle if the refreshInterval/heartbeatInterval is undefined or not. Also, there is a limit in the config schema for both of these fields. If a user sets the value above the maximum by editing the raw config.json, then it's their responsibility in my opinion.

kovapatrik replied:
I might be nitpicking here a bit :D but don't want to have unneccessary lines in the code.

dkerr64 replied:
Yes, you're right. Will fix next time I commit.

kovapatrik commented on src/platform.ts:
```diff
@@ -79,10 +92,46 @@ export class MideaPlatform implements DynamicPlatformPlugin {
+      setTimeout(() => {
+        missingDevices = 0;
+        // on repeat attempts only retry once (so two broadcasts sent), 3000 miliseconds apart.
+        this.discover.startDiscover(1, 3000);
```
Maybe I would set the timeout to even bigger, like every 5 minutes, although I don't know how heavy the operation to send discovery messages on all interfaces.

dkerr64 replied:
I wondered about that.  But I think that a broadcast once a minute is pretty lightweight on a LAN.  I pondered making it configurable, but you have to draw the line somewhere in what makes sense to allow a user to tinker with.  Happy to change it to a longer duration if you think it better.

kovapatrik commented on src/platform.ts:
```diff
@@ -134,9 +183,8 @@ export class MideaPlatform implements DynamicPlatformPlugin {
           if (deviceConfig.advanced_options.token && deviceConfig.advanced_options.key) {
+            // token/key provided in config file, use those... set values in cached context
             this.log.info(`[${device_info.name}] New device at ${device_info.ip}:${device_info.port}, using token/key from config file`);
             accessory.context.token = deviceConfig.advanced_options.token;
```
When we began to refactor the code to the new structure (where we have the token/key in the config.json) I thought about if we need to store them in the cache or not as they will be available all the time through the config file. What are your thoughts about this?

dkerr64 replied:
Now that we **require** a devices array, I think it probably is redundant to cache.  When would it ever be used?  Do we have code to remove a cached accessory if it is removed from config.json?

For that matter... we don't have a button in the settings GUI to delete a device.  Oops.  Need to add that.

dkerr64 replied:
Mind you, we should be cautious about deleting a cached homebridge accessory.  If the device is removed from the config.json just temporarily, then is put back, what is the user expectation?  If we delete the accessory from cache then I think Apple Homekit is going to delete it.... along with any automations connected to it.  And that is nasty. I have had to rebuild a bunch of automations when I tinkered with some Homebridge accessories, it is not fun... and, Apple in their wisdom don't provide a backup/restore for Home automations. What are they thinking?

kovapatrik replied:
There is a delete button in the GUI by default. It's a little 'x' in the top right corner, when you set the fields for a device.

Currently we don't delete accessories at all I think, cached devices can be deleted by hand using the GUI. I think it should remain as is.

dkerr64 replied:
You're right, delete 'X' is there.

Right now if you delete a device, but it is still on your network, then on restart the plugin prints a warning in the log that a new device is discovered and that you should add it to config.json.  `AddDevice()` is NOT called, so the cached accessory remains visible in Homebridge, but is inactive.

We could leave it as-is.  Or we could delete the cached accessory.  Or we could call AddDevice... essentialy ignoring the removal from config.json.

If we leave it as-is, then I think we need to document that the user must manually delete the cached accessory in Homebridge.  Deleting the cached accessory in the plugin is bad for reason I stated above.  And I don't think we should simply ignore a user deleting it.  So I think best is to leave the code as-is and update the documentation.

kovapatrik replied:
All of the plugins I use applied the same methodology that we have talked about, so I think this is the best way. In the other hand we don't use the cache for anything (I think?) so it could be safely deleted.

kovapatrik replied:
No, I'm wrong, we can't delete the cache for one reason: it contains the UUID, which are being used by Homekit I guess. So yeah, as you said, if we delete it, and later we add it back, the automations won't contain the readded device.

dkerr64 replied:
I updated the readme, and fixed issue in index.html where the table header titles could be split onto multiple lines when they shouldn't be.  And an we were setting serial number and model into the accessory cache for devices already cached instead of newly added devices!

dkerr64 commented:
The last thing bugging me is Single Accessory setting.  What do you want to do with it?

kovapatrik commented:
It can be deleted. If we don't use it, we will mimic the default behaviour of a Homekit device.

dkerr64 commented:
> It can be deleted. If we don't use it, we will mimic the default behaviour of a Homekit device.

Is the default behavior equivalent to single device = true?

kovapatrik commented:
Yes, it is.

dkerr64 commented:
> Yes, it is.

Done.

kovapatrik reviewed (approved):
I don't have anything in my mind which we should change. If you agree, I will merge.

kovapatrik commented:
Maybe one thing: what version should we bump to? Can we call this as 1.0.0? If we do that, maybe we can also start the official verification of the plugin.

dkerr64 commented:
> Maybe one thing: what version should we bump to? Can we call this 1.0.0? If we do that, maybe we can also start the official verification of the plugin.

Well, that is quite a milestone! My only reservation is that we have at least one user reporting protocol errors.  I think perhaps we merge without bumping to 1.0.0 and ask him to test.  Let's see if we can clean up any of those errors.  I'd also like some review of the README... get a sense if the documentation is clear enough.  If nothing major comes up then bump the version number.

kovapatrik commented:
Sure thing, but we must bump the version to something else. npm doesn't allow to publish with an existing version number.

dkerr64 commented:
> Sure thing, but we must bump the version to something else. npm doesn't allow to publish with an existing version number.

Yes of course.  0.4.0 maybe !

kovapatrik commented:
Done! Can I merge then?

dkerr64 commented:
> Done! Can I merge then?

Yes, let's merge.  Exciting.  !!

kovapatrik commented:
I've added issue #19 to this PR, which I thought will be resolved without testing. Did I do that right?

dkerr64 commented:
> I've added issue #19 to this PR, which I thought will be resolved without testing. Did I do that right?

It would be normal to get verification from person that opened the issue that it was fixed before closing it.  I juts posted a comment in there asking him to test.
